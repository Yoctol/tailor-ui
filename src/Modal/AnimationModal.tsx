import React, { FC, KeyboardEventHandler, MouseEventHandler } from 'react';
import { animated, config, useSpring, useTransition } from '@react-spring/web';

import { Backdrop } from '../Backdrop';
import { ESC_KEY_CODE, useKeydown } from '../hooks';
import { Portal } from '../Portal';
import { StackingOrder } from '../constants';
import { StatusType } from '../types';

import { ModalContent, ModalStatusBar, ModalWrapper } from './styles';
import { ModalSize } from './types';

const AnimatedModalWrapper = animated(ModalWrapper);
const AnimatedModalStatusBar = animated(ModalStatusBar);

export interface AnimationModalProps {
  onCancel: MouseEventHandler | KeyboardEventHandler;
  onOpenComplete?: () => void;
  onCloseComplete?: () => void;
  size?: ModalSize;
  closable?: boolean;
  status?: StatusType | null;
  visible: boolean;
  zIndex?: number;
}

const AnimationModal: FC<AnimationModalProps> = ({
  children = '',
  visible,
  onCancel,
  onOpenComplete,
  onCloseComplete,
  closable = true,
  size = 'md',
  status = null,
  zIndex = StackingOrder.OVERLAY,
  ...otherProps
}) => {
  useKeydown({
    listening: closable ? visible : false,
    keyCode: ESC_KEY_CODE,
    onKeydown: onCancel as any,
  });

  const transitions = useTransition(visible, {
    from: {
      opacity: 0,
      transform: 'translate(-50%, -50%) scale(0.9)',
      pointerEvents: 'auto',
    },
    enter: {
      opacity: 1,
      transform: 'translate(-50%, -50%) scale(1)',
      pointerEvents: 'auto',
    },
    leave: {
      opacity: 0,
      transform: 'translate(-50%, -50%) scale(0.9)',
      pointerEvents: 'none',
    },
    onDestroyed: (isDestroyed) => {
      if (isDestroyed && onCloseComplete && !visible) {
        onCloseComplete();
      } else if (!isDestroyed && onOpenComplete && visible) {
        onOpenComplete();
      }
    },
    config: config.stiff,
  });

  const statusProps = useSpring({
    opacity: visible ? 1 : 0,
    top: visible ? -10 : 0,
  });

  return (
    <>
      <Backdrop
        visible={visible}
        zIndex={zIndex}
        onClick={(event) => {
          if (closable) {
            onCancel(event as any);
          }
        }}
      />
      {transitions(
        (style, item) =>
          item && (
            <Portal defaultOrder={zIndex}>
              <AnimatedModalWrapper style={style}>
                {status && (
                  <AnimatedModalStatusBar status={status} style={statusProps} />
                )}
                <ModalContent size={size} {...otherProps}>
                  {children}
                </ModalContent>
              </AnimatedModalWrapper>
            </Portal>
          )
      )}
    </>
  );
};

export default AnimationModal;
