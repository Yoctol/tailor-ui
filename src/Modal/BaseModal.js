import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import styled from 'styled-components';
import { Transition, animated, config } from 'react-spring';
import { space, themeGet, width } from 'styled-system';

import Keydown from '../utils/Keydown';

const ModalOverlay = styled.div`
  position: fixed;
  z-index: 10000;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.65);
`;

const AnimatedModalOverlay = animated(ModalOverlay);

const ModalContent = styled.div`
  display: flex;
  position: fixed;
  z-index: 10001;
  top: 50%;
  left: 50%;
  flex-direction: column;
  max-height: 90vh;
  border-radius: ${themeGet('radii.base')};
  background-color: #fff;
  box-shadow: 0 10px 30px 0 rgba(17, 17, 17, 0.2);
  transform: translate(-50%, -50%);

  ${width};
`;

ModalContent.propTypes = {
  ...space.propTypes,
  ...width.propTypes,
};

ModalContent.defaultProps = {
  width: 416,
};

const AnimatedModalContent = animated(ModalContent);

const ModalWrapper = ({
  opacity,
  translateY,
  pointerEvents,
  onCancel,
  content,
  clickOutsite,
  ...otherProps
}) => (
  <>
    <AnimatedModalOverlay
      style={{
        opacity,
        pointerEvents,
      }}
      onClick={() => {
        if (clickOutsite) {
          onCancel();
        }
      }}
    />
    <AnimatedModalContent
      style={{
        opacity,
        pointerEvents,
      }}
      {...otherProps}
    >
      {content}
    </AnimatedModalContent>
  </>
);

const ESC_KEY_CODE = 27;

class BaseModal extends PureComponent {
  render() {
    const { children, visible, onCancel, ...otherProps } = this.props;

    return (
      <>
        {visible && <Keydown keyCode={ESC_KEY_CODE} handleKeydown={onCancel} />}
        <Transition
          native
          keys={visible ? 'visible' : 'hide'}
          from={{
            opacity: 0,
          }}
          enter={{
            opacity: 1,
          }}
          leave={{
            opacity: 0,
            pointerEvents: 'none',
          }}
          config={config.stiff}
          onCancel={onCancel}
          content={children}
          {...otherProps}
        >
          {visible ? ModalWrapper : null}
        </Transition>
      </>
    );
  }
}

BaseModal.propTypes = {
  children: PropTypes.node,
  clickOutsite: PropTypes.bool,
  visible: PropTypes.bool.isRequired,
  onCancel: PropTypes.func.isRequired,
};

BaseModal.defaultProps = {
  children: null,
  clickOutsite: true,
};

export default BaseModal;
