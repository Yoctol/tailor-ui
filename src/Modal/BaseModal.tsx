import React, { PureComponent, ReactNode, SFC } from 'react';
import styled from 'styled-components';
import { Transition, animated, config } from 'react-spring';
import { WidthProps, width as styledWidth, themeGet } from 'styled-system';

import Keydown from '../utils/Keydown';

const ModalOverlay = styled.div`
  position: fixed;
  z-index: 10000;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.65);
`;

const AnimatedModalOverlay = animated(ModalOverlay);

const ModalContent = styled<WidthProps, 'div'>('div')`
  display: flex;
  position: fixed;
  z-index: 10001;
  top: 50%;
  left: 50%;
  flex-direction: column;
  max-height: 90vh;
  border-radius: ${themeGet('radii.base')};
  background-color: #fff;
  box-shadow: 0 10px 30px 0 rgba(17, 17, 17, 0.2);
  transform: translate(-50%, -50%);

  ${styledWidth};
`;

const AnimatedModalContent = animated(ModalContent);

export type ModalWrapperProps = WidthProps & {
  opacity: number;
  pointerEvents?: string;
  onCancel: () => void;
  content?: ReactNode;
  clickOutsite?: boolean;
  [key: string]: any;
};

const ModalWrapper: SFC<ModalWrapperProps> = ({
  width = 416,
  opacity,
  pointerEvents,
  onCancel,
  content = '',
  clickOutsite = true,
  ...otherProps
}) => (
  <>
    <AnimatedModalOverlay
      style={{
        opacity,
        pointerEvents,
      }}
      onClick={() => {
        if (clickOutsite) {
          onCancel();
        }
      }}
    />
    <AnimatedModalContent
      style={{
        opacity,
        pointerEvents,
      }}
      width={width}
      {...otherProps}
    >
      {content}
    </AnimatedModalContent>
  </>
);

const ESC_KEY_CODE = 27;

export type BaseModalProps = WidthProps & {
  onCancel: () => void;
  clickOutsite?: boolean;
  visible: boolean;
};

class BaseModal extends PureComponent<BaseModalProps> {
  render() {
    const { children, visible, onCancel, ...otherProps } = this.props;

    return (
      <>
        {visible && <Keydown keyCode={ESC_KEY_CODE} handleKeydown={onCancel} />}
        <Transition
          native
          keys={visible ? 'visible' : 'hide'}
          from={{
            opacity: 0,
          }}
          enter={{
            opacity: 1,
          }}
          leave={{
            opacity: 0,
            pointerEvents: 'none',
          }}
          config={config.stiff}
          onCancel={onCancel}
          content={children}
          {...otherProps}
        >
          {visible && ModalWrapper}
        </Transition>
      </>
    );
  }
}

export default BaseModal;
