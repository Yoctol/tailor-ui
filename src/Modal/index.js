import React from 'react';
import PropTypes from 'prop-types';

import ModalWithAnimation from './ModalWithAnimation';
import ModalWithoutAnimation from './ModalWithoutAnimation';

const Modal = ({ animation = false, ...otherProps }) => {
  const BaseModal = animation ? ModalWithAnimation : ModalWithoutAnimation;
  return <BaseModal {...otherProps} />;
};

Modal.propTypes = {
  animation: PropTypes.bool,
};

Modal.defaultProps = {
  animation: false,
};

export default Modal;
