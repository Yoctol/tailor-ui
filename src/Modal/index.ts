export * from './Modal';
export { useModal } from './useModal';
