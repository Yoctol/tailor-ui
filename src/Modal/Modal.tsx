import React, { CSSProperties, FC } from 'react';

import { Flex } from '../Layout';

import AnimationModal, { AnimationModalProps } from './AnimationModal';
import ModalFooter, { ModalFooterProps } from './Footer';
import ModalHeader, { ModalHeaderProps } from './Header';

export type ModalProps = AnimationModalProps &
  ModalHeaderProps &
  ModalFooterProps & {
    zIndex?: number;
    contentStyle?: CSSProperties;
    footerStyle?: CSSProperties;
  };

const Modal: FC<ModalProps> = ({
  title,
  status,
  onCancel,
  onConfirm,
  closable,
  children,
  footer,
  cancelText,
  confirmText,
  confirmButtonProps,
  cancelButtonProps,
  zIndex,
  contentStyle,
  footerStyle,
  ...props
}) => (
  <AnimationModal
    status={status}
    onCancel={onCancel}
    zIndex={zIndex}
    {...props}
  >
    <ModalHeader
      status={status}
      title={title}
      onCancel={onCancel}
      closable={closable}
    />
    <Flex
      flex="auto"
      flexDirection="column"
      overflowY="auto"
      mx="3"
      mb="24px"
      px="3"
      style={contentStyle}
    >
      {children}
    </Flex>
    <ModalFooter
      footer={footer}
      closable={closable}
      cancelText={cancelText}
      confirmText={confirmText}
      onCancel={onCancel}
      onConfirm={onConfirm}
      confirmButtonProps={confirmButtonProps}
      cancelButtonProps={cancelButtonProps}
      style={footerStyle}
    />
  </AnimationModal>
);

export { Modal };
