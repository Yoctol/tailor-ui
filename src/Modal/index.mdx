---
name: Modal
menu: Components
route: /components/modal
---

import { Toggle, State } from 'react-powerplug';
import { PropsTable, Playground } from 'docz';
import Modal from './';
import { Button } from '../';

# Modal

Modal dialogs.

## When To Use

When requiring users to interact with the application, but without jumping to a new page and interrupting the user's workflow, you can use Modal to create a new floating layer over the current page to get user feedback or display information.

## Examples

```js
import { Modal } from '@yoctol/ui';
```

### Basic

<Playground>
  <Toggle>
    {({ on, toggle }) => (
      <div>
        <Modal
          title="This is a Modal"
          visible={on}
          closable
          onConfirm={toggle}
          onCancel={toggle}
        >
          {Array.from({ length: 100 }).map((_, index) => (
            <div key={index}>This is the content of Modal</div>
          ))}
        </Modal>
        <Button onClick={toggle}>Open Modal</Button>
      </div>
    )}
  </Toggle>
</Playground>

### Customize footer

<Playground>
  <Toggle>
    {({ on, toggle }) => (
      <div>
        <Modal
          title="This is a Modal"
          visible={on}
          footer={<div>My footer</div>}
          closable
          onCancel={toggle}
        >
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
        </Modal>
        <Button onClick={toggle}>Open Modal</Button>
      </div>
    )}
  </Toggle>
</Playground>

### Without footer

<Playground>
  <Toggle>
    {({ on, toggle }) => (
      <div>
        <Modal
          title="This is a Modal"
          visible={on}
          footer={null}
          closable
          onCancel={toggle}
        >
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
        </Modal>
        <Button onClick={toggle}>Open Modal</Button>
      </div>
    )}
  </Toggle>
</Playground>

### With button props

<Playground>
  <State initial={{ visible: false, loading: false }}>
    {({ state, setState }) => (
      <div>
        <Modal
          title="This is a Modal"
          visible={state.visible}
          closable
          onConfirm={() => {
            setState({ loading: true });
            setTimeout(() => {
              setState({ visible: false, loading: false });
            }, 3000);
          }}
          onCancel={() => setState({ visible: false })}
          confirmText="DELETE"
          confirmButtonProps={{ type: 'danger', loading: state.loading }}
        >
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
          <div>This is the content of Modal</div>
        </Modal>
        <Button onClick={() => setState({ visible: true })}>Open Modal</Button>
      </div>
    )}
  </State>
</Playground>

### Instance

#### Confirm

##### Using callback

<Playground>
  <Button
    onClick={() => {
      Modal.confirm({
        title: 'Do you Want to delete these items?',
        content: 'Some descriptions',
        onConfirm: () => console.log('confirmed!'),
        onCancel: () => console.log('canceled!'),
      });
    }}
  >
    Confirm
  </Button>
</Playground>

##### Using promise

<Playground>
  <Button
    onClick={async function() {
      const confirmed = await Modal.confirm({
        title: 'Do you want to delete these items?',
        content: 'Some descriptions',
      });

      if (confirmed) {
        console.log('confirmed!');
      } else {
        console.log('canceled!');
      }
    }}>
    Confirm

  </Button>
</Playground>

#### Status

<Playground>
  <Button
    onClick={() => {
      Modal.info({
        title: 'This is a info modal',
        content: 'Some descriptions',
        onConfirm: () => console.log('info confirm'),
      })
    }}
  >
    Info
  </Button>
  <Button
    ml="2"
    onClick={async function() {
      const confirmed = await Modal.success({
        title: 'This is a success modal',
        content: 'Some descriptions',
      });

      if (confirmed) {
        console.log('success confirm');
      }
    }}>
    Success

  </Button>
  <Button
    ml="2"
    onClick={() => {
      Modal.warning({
        title: 'This is a warning modal',
        content: 'Some descriptions',
      })
    }}
  >
    Warning
  </Button>
  <Button
    ml="2"
    onClick={() => {
      Modal.error({
        title: 'This is a error modal',
        content: 'Some descriptions',
      })
    }}
  >
    Error
  </Button>
</Playground>

## API

### Modal

<PropsTable of={Modal} />

### Instance

- `Modal.confirm(options) => Promise`
- `Modal.info(options) => Promise`
- `Modal.success(options) => Promise`
- `Modal.warning(options) => Promise`
- `Modal.error(options) => Promise`

### `options`

| Property    | Type                | Default | Description                                                                                                                                                                    |
| ----------- | ------------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| cancelText  | string              | Cancel  | Text of the Cancel button                                                                                                                                                      |
| confirmText | string              | Conrim  | Text of the Conrim button                                                                                                                                                      |
| onCancel    | function            | -       | Specify a function that will be called when the user clicks the Cancel button. The parameter of this function is a function whose execution should include closing the dialog. |
| onConfirm   | function            | -       | Specify a function that will be called when the user clicks the OK button. The parameter of this function is a function whose execution should include closing the dialog.     |
| title       | string \| ReactNode | -       | Title                                                                                                                                                                          |
| content     | string \| ReactNode | -       | content of the message                                                                                                                                                         |
