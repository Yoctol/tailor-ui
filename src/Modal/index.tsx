import React, {
  FunctionComponent,
  ReactNode,
  isValidElement,
  useContext,
} from 'react';

import Box from '../Grid/Box';
import Flex from '../Grid/Flex';
import Heading from '../Heading';
import Portal from '../utils/Portal';
import { LocaleContext } from '../UIProvider';

import BaseModal, { BaseModalProps } from './BaseModal';
import CloseButton from './CloseButton';
import Footer, { IFooterProps } from './Footer';
import { confirm, error, info, success, warning } from './instance';

interface IModalHeaderProps {
  title?: ReactNode;
  closable?: boolean;
  onCancel: () => void;
}

export const ModalHeader: FunctionComponent<IModalHeaderProps> = ({
  title,
  onCancel,
  closable,
}) => (
  <Flex
    flex="none"
    alignItems="center"
    px="3"
    py="2"
    borderBottom="base"
    borderColor="gray300"
  >
    <Box flex="auto">
      <Heading.h3>{title}</Heading.h3>
    </Box>
    {closable && <CloseButton onCancel={onCancel} />}
  </Flex>
);

export const ModalContent: FunctionComponent = ({ children }) => (
  <Flex flex="auto" flexDirection="column" p="3" overflowY="auto">
    {children}
  </Flex>
);

export const FooterWrapper: FunctionComponent = props => (
  <Flex
    flex="none"
    alignItems="center"
    px="3"
    py="2"
    borderTop="base"
    borderColor="gray300"
    {...props}
  />
);

interface IModalFooterProps extends IFooterProps {
  footer?: ReactNode;
}

const ModalFooter: FunctionComponent<IModalFooterProps> = ({
  footer,
  cancelText,
  confirmText,
  onCancel,
  onConfirm,
  confirmButtonProps,
  cancelButtonProps,
}) => {
  const { locale } = useContext(LocaleContext);

  return footer === null ? null : (
    <FooterWrapper>
      {isValidElement(footer) ? (
        footer
      ) : (
        <Footer
          cancelText={cancelText || locale.Modal.cancelText}
          confirmText={confirmText || locale.Modal.confirmText}
          onCancel={onCancel}
          onConfirm={onConfirm}
          confirmButtonProps={confirmButtonProps}
          cancelButtonProps={cancelButtonProps}
        />
      )}
    </FooterWrapper>
  );
};

export type ModalProps = BaseModalProps &
  IFooterProps &
  IModalHeaderProps &
  IModalFooterProps;

const Modal: FunctionComponent<ModalProps> & {
  confirm: typeof confirm;
  info: typeof info;
  success: typeof success;
  warning: typeof warning;
  error: typeof error;
} = ({
  title,
  onCancel,
  onConfirm,
  closable,
  children,
  footer,
  cancelText,
  confirmText,
  confirmButtonProps,
  cancelButtonProps,
  ...props
}) => (
  <Portal>
    <BaseModal onCancel={onCancel} {...props}>
      <ModalHeader title={title} onCancel={onCancel} closable={closable} />
      <ModalContent>{children}</ModalContent>
      <ModalFooter
        footer={footer}
        cancelText={cancelText}
        confirmText={confirmText}
        onCancel={onCancel}
        onConfirm={onConfirm}
        confirmButtonProps={confirmButtonProps}
        cancelButtonProps={cancelButtonProps}
      />
    </BaseModal>
  </Portal>
);

Modal.confirm = confirm;
Modal.info = info;
Modal.success = success;
Modal.warning = warning;
Modal.error = error;

export default Modal;
