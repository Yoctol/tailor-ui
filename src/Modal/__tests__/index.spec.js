import React from 'react';

import shallowWithTheme from 'test/shallowWithTheme';

import Modal from '..';

describe('Modal', () => {
  it('should render correctly', () => {
    const component = shallowWithTheme(
      <Modal visible handleClose={() => {}} />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render with props closeButton', () => {
    const component = shallowWithTheme(
      <Modal visible closable handleClose={() => {}} />
    );

    expect(component).toMatchSnapshot();
  });
});
