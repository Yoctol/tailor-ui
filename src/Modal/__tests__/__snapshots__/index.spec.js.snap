// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render correctly 1`] = `
<Modal
  closeButton={false}
  handleClose={[Function]}
  show={true}
/>
`;

exports[`Modal should render with props closeButton 1`] = `
<Modal
  closeButton={true}
  handleClose={[Function]}
  show={true}
/>
`;
